### ISSUE TITLE
[Feature Request] Fused Linear Cross-Entropy API in tf.nn for memory-efficient LLM training

### ISSUE THREAD

**Issue by powerofaisinstudy-debug** [tensorflow/tensorflow#127290]
### Issue type

Feature Request

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.x / nightly

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.10 or 3.11

### Current behavior?

Standard tf.matmul followed by tf.nn.sparse_softmax_cross_entropy_with_logits requires writing a massive intermediate [Batch, SeqLen, Vocab] logit tensor to global GPU VRAM. For a batch size of 16, sequence length of 4096, and vocabulary size of 128,000, storing this intermediate tensor requires ~16.7 GB of VRAM. This causes severe GPU bandwidth bottlenecks and frequent dynamic memory allocation Out-Of-Memory (OOM) errors during large-scale model training.

Expected Behavior:
A native fused operator (tf.nn.fused_linear_cross_entropy) that computes the linear output projection and cross-entropy loss in a single GPU pass using online log-sum-exp reductions in SRAM, bypassing intermediate VRAM logit allocations entirely.

### Proposed API Design
```python
tf.nn.fused_linear_cross_entropy(
    labels=labels,
    inputs=hidden_states,
    weight=lm_head_weight,
    bias=bias,
    ignore_index=-100,
    label_smoothing=0.0,
    reduction="weighted_mean"
)

Key BenefitsZero Logit VRAM Allocation: Drops intermediate memory overhead from gigabytes to $O(1)$ relative to vocabulary size.Bandwidth Savings: Eliminates redundant global memory read/write cycles between tf.matmul and tf.nn.softmax.Numerical Stability: Uses chunked online softmax reductions inside GPU shared memory.Alternatives ConsideredManual Python chunking loops incur severe tf.function tracing overhead and high CUDA kernel launch penalties. Bringing a native, XLA-compatible fused cross-entropy kernel to tf.nn provides parity with modern high-performance LLM training tools (e.g., Liger Kernel).

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

# Simulating large-vocabulary LLM output projection
batch_size = 16
seq_len = 4096
vocab_size = 128000
hidden_dim = 4096

hidden_states = tf.random.normal([batch_size, seq_len, hidden_dim], dtype=tf.bfloat16)
lm_head_weight = tf.random.normal([vocab_size, hidden_dim], dtype=tf.bfloat16)
labels = tf.random.uniform([batch_size, seq_len], maxval=vocab_size, dtype=tf.int32)

# Unfused operation: Allocates ~16.7 GB intermediate logit matrix
logits = tf.matmul(hidden_states, lm_head_weight, transpose_b=True)
loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits)

print("Loss computed successfully, but required allocating intermediate logits tensor.")
```

### Relevant log output

```shell

```

**Comment by Kayyuri** [comment]
Hi @powerofaisinstudy-debug, Thanks for your interest in contributing.

You’ve clearly demonstrated the value of a `tf.nn.fused_linear_cross_entropy`, which could significantly reduce memory usage and bandwidth overhead during large-vocabulary LLM training.

If you decide to proceed, please submit a pull request in accordance with the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute) and link this issue in your PR so the maintainers can review it and provide feedback.

Thank you!

**Comment by Kayyuri** [comment]
Hi @powerofaisinstudy-debug, 

Thank you for raising the [PR](https://github.com/tensorflow/tensorflow/pull/127422). We truly appreciate your initiative and the effort in contributing the fix to Tensorflow.

**Comment by powerofaisinstudy-debug** [comment]
Thank you so much @Kayyuri! Glad I could contribute to TensorFlow. Looking forward to seeing it merged!

**Comment by OssAgenticsT** [comment]
Closing this issue as resolved. Feel free to open a new issue or comment here if you need further assistance!
